CMakeIntegratedOpenCL.cmake

```diff
@@ -0,0 +1,88 @@
+set(OpenCL_INCLUDE_DIR ${opencl-headers_SOURCE_DIR} CACHE PATH "") # for Boost::Compute
+
+FetchContent_Declare(OpenCL-ICD-Loader GIT_REPOSITORY ${OPENCL_LOADER_REPOSITORY} GIT_TAG ${OPENCL_LOADER_TAG})
+FetchContent_GetProperties(OpenCL-ICD-Loader)
+if(NOT OpenCL-ICD-Loader_POPULATED)
+  FetchContent_Populate(OpenCL-ICD-Loader)
+  set(USE_DYNAMIC_VCXX_RUNTIME ON)
+  add_subdirectory(${opencl-icd-loader_SOURCE_DIR} ${opencl-icd-loader_BINARY_DIR} EXCLUDE_FROM_ALL)
+  message(STATUS "Populated OpenCL ICD Loader")
+endif()
+list(APPEND INTEGRATED_OPENCL_INCLUDES ${OPENCL_ICD_LOADER_HEADERS_DIR})
+list(APPEND INTEGRATED_OPENCL_LIBRARIES ${opencl-icd-loader_BINARY_DIR}/Release/OpenCL.lib cfgmgr32.lib runtimeobject.lib)
+list(APPEND INTEGRATED_OPENCL_DEFINITIONS CL_TARGET_OPENCL_VERSION=120)
+
+# Build Independent Boost libraries
+include(ExternalProject)
+include(ProcessorCount)
+ProcessorCount(J)
+set(BOOST_BASE "${PROJECT_BINARY_DIR}/Boost")
+set(BOOST_BOOTSTRAP "${BOOST_BASE}/source/bootstrap.bat")
+set(BOOST_BUILD "${BOOST_BASE}/source/b2.exe")
+set(BOOST_FLAGS "")
+list(APPEND BOOST_SUBMODULES "libs/*" "tools/*")
+ExternalProject_Add(Boost
+  TMP_DIR "${BOOST_BASE}/tmp"
+  STAMP_DIR "${BOOST_BASE}/stamp"
+  DOWNLOAD_DIR "${BOOST_BASE}/download"
+  SOURCE_DIR "${BOOST_BASE}/source"
+  BINARY_DIR "${BOOST_BASE}/source"
+  INSTALL_DIR "${BOOST_BASE}/install"
+  GIT_REPOSITORY ${BOOST_REPOSITORY}
+  GIT_TAG ${BOOST_TAG}
+  GIT_SUBMODULES ${BOOST_SUBMODULES}
+  GIT_SHALLOW ON
+  UPDATE_COMMAND ""
+  PATCH_COMMAND ""
+  CONFIGURE_COMMAND ${BOOST_BOOTSTRAP}
+  BUILD_COMMAND ${BOOST_BUILD} -sBOOST_ROOT=${BOOST_BASE}/source -a -q -j ${J} --with-headers --with-chrono --with-filesystem --with-system link=static runtime-link=shared variant=release threading=multi cxxflags="${BOOST_FLAGS}"
+  INSTALL_COMMAND ""
+)
+set(BOOST_INCLUDE ${BOOST_BASE}/source CACHE PATH "")
+set(BOOST_LIBRARY ${BOOST_BASE}/source/stage/lib CACHE PATH "")
+list(APPEND INTEGRATED_OPENCL_INCLUDES ${BOOST_INCLUDE})
+if(MSVC)
+  if(${MSVC_VERSION} GREATER 1929)
+    message(FATAL_ERROR "Unrecognized MSVC version number")
+  elseif(${MSVC_VERSION} GREATER 1919)
+    set(MSVC_TOOLCHAIN_ID "142")
+  elseif(${MSVC_VERSION} GREATER 1909)
+    set(MSVC_TOOLCHAIN_ID "141")
+  elseif(${MSVC_VERSION} GREATER 1899)
+    set(MSVC_TOOLCHAIN_ID "140")
+  else()
+    message(FATAL_ERROR "Unrecognized MSVC version number")
+  endif()
+  list(APPEND INTEGRATED_OPENCL_LIBRARIES ${BOOST_LIBRARY}/libboost_filesystem-vc${MSVC_TOOLCHAIN_ID}-mt-x64-${BOOST_VERSION_UNDERSCORE}.lib)
+  list(APPEND INTEGRATED_OPENCL_LIBRARIES ${BOOST_LIBRARY}/libboost_system-vc${MSVC_TOOLCHAIN_ID}-mt-x64-${BOOST_VERSION_UNDERSCORE}.lib)
+  list(APPEND INTEGRATED_OPENCL_LIBRARIES ${BOOST_LIBRARY}/libboost_chrono-vc${MSVC_TOOLCHAIN_ID}-mt-x64-${BOOST_VERSION_UNDERSCORE}.lib)
+else()
+  message(FATAL_ERROR "MinGW Boost library names not yet specified")
```

Comment:
This is a terrible message, I'm sorry for that -- I'm pretty sure it was just a placeholder I forgot to change.

The underlying issue that because MinGW installations are so varied, it's more difficult to predict the name and location of the .lib files.  So this should be addressed in a separate PR.

I've changed the message to show that MinGW is not supported yet.

Reply:
Got it! Thank you!

I'm not sure but it seems that it even doesn't worth to implement MinGW support. Anyway, we will create nightly and release wheels under VS for the performance purposes (https://lightgbm.readthedocs.io/en/latest/FAQ.html#i-am-using-windows-should-i-use-visual-studio-or-mingw-for-compiling-lightgbm). As long as compiled with VS library doesn't conflict with MinGW (have no idea how can it happen) and this option remains "private" (for repo maintainers or very advanced third-party developers), MinGW support is not needed, I believe. Please correct me if I'm wrong.

Reply:
I'm far from an expert, but my impression is also that MinGW support is significantly less urgent since MSFT began making the MSVC community edition available for free.